index.bs

```diff
@@ -375,7 +375,81 @@ request they are associated with.
+    greater than or equal to 14 (to avoid being too weak) and less than or equal
+    to 128 (the limit defined by scrypt).  A value of 15 is recommended (an
+    scrypt N of 2^15 or 32768).
+
+The responder replies with an authentication-response message with the following values:
+
+-   result:  If the responder was able to calculate proof of possession of the
+    shared secret, and if it failed, why it failed.
+
+-   proof: The result of running
+    [HKDF](https://tools.ietf.org/html/rfc5869) on the result of running
+    [scrypt](https://tools.ietf.org/html/rfc7914) on the shared secret.  For
+    scrypt, the cost parameter (N) used is the 2 to the power of the cost from
+    the authentication-request message, the block size (r) is 8, the
+    parallelization parameter (p) is 1, and the key length derived is 32.  For
+    HKDF, both the extract and expand steps are used, the hash function used is
+    sha-256, the salt used is the salt from the authentication-request message,
+    the application-specific info is a serialized certificate-fingerprint-pair
+    object (defined below), and the length of the output key matrial (L) is 32.
+
+Note: the values of 32 above (for salt length, scrypt output key length, and
+HKDF output key material length) are based on the output size of sha-256.  If a
+different hash mechanism is used in the future, these values should be updated as
+well.
+
+The certificate-fingerprint-pair object used as application-specific info above
+is serialized as CBOR (without the type or length prefix of messages in a QUIC
+stream) and contains the following values:
+
+-   challenger-fingerprint: The certificate fingerprint of the
+    challenger, used in the QUIC crypto handshake during connection
+    establishment, as defined by [RFC 8122 section
+    5](https://tools.ietf.org/html/rfc8122#section-5).  It is the output using
+    sha-256 as the hash wihtout any hexadecimal encoding or prefixing.
+
+-   responder-fingerprint: The certificate fingerprint of the
+    responder, obtained from the QUIC crypto handshake during connection
+    establishment, as defined by [RFC 8122 section
+    5](https://tools.ietf.org/html/rfc8122#section-5).  It is the output using
+    sha-256 as the hash wihtout any hexadecimal encoding or prefixing. 
+
+To verify that the responder's proof is correct, the challenger makes the same
+calculation of the proof and compares the result. If the results are the same,
+the challenger considers the responder authenticated, and considers it
+unauthenticated otherwise.  In either case, the challenger sends the responder an
+authentication-result message with the following values:
+
+-   result:  If the challenger was able to authenticate the responder or not,
+    and if not, why not.
+
+The challenger must limit the time the responder has to send a response to 60
+seconds (to avoid the possibility of brute-force attacks.)
```

Comment:
It's probably a good idea to rate limit the challenger, since it may be possible to force a challenge to be sent through the drive-by web.   However, to prevent brute force attacks on the responder, the responder should also be rate limited.

Reply:
Why rate-limit a challenger?  The challenger isn't the one authenticating; the responder is.  There would be no reason to challenge many times, unless there's some way to learn the pre-shared secret using a variety of salts.  Is that what you're worried about?

The responder is rate limited here already in two ways: 1 by scrypt.  It can only compute so many responses.   2 by the challenger which will only issues so many challenges (probably one per QUIC connection).  

Or perhaps that's what you meant by rate limit the challenger?  Only attempt N (perhaps = 1) times per QUIC connection?

Reply:
If an application acting as a challenger can force a challenge to be sent to a specific responder, it can create a denial of service to the responder.   Most reasonable implementations should mitigate this by aggregating connections across applications/sites, so an application can't force multiple connections to be attempted to the same receiver.  However, rate limiting in the challenger implementation would be an added layer of protection.

Since I wasn't thinking of crypto vulnerabilities, it feels more like  a security mitigation in the implementation, instead of a spec issue for the authentication handshake.  In which case I'm happy to discuss rate limiting in my other pull request - no changes required here.